### DISCUSSION TITLE
yourTTS streaming?

### DISCUSSION THREAD

**Discussion by Disastorm** [coqui-ai/TTS#1764] [General Q&A] [answered]
Hello.

I've made an application that essentially streams audio from an input in chunks into modified versions of the transfer_voice and tts functions from the coqui-ai TTS repository files using the yourTTS model.

However at the area where the chunks connect, they don't continue cleanly (after conversion), I guess because each chunk doesn't have the data from the previous one to continue the audio smoothly.  Is there a way to solve this? 

I don't actually know much about AI or the tensor libraries themselves, I just did this using modified versions of the existing functions in the coquiTTS utils files.

I did try saving the audio of the previous chunk, prepending it to the current audio and pushing it through the tts functions and then cutting the output in half, but unless I did it wrong, it seems that method, while it does sound better, still doesn't sound completely smooth either, so unless I did it wrong, I'm guessing thats not the correct solution either.

**Comment by Disastorm** [comment, marked as answer]
In case anyone is wondering, I actually ended up getting it to sound pretty decent by expanding my above idea and sending 3 chunks each time and extracting just the output for the middle chunk.

To generate chunk 2's audio you send chunk 1, 2, and 3.
To generate chunk 3's audio you send chunk 2, 3, and 4.
Although of course this means the output gets delayed by an additional chunk length.

I don't know if maybe there is an actual function somewhere that allows the generation of chunk output but continuing from the previous chunk smoothly, but the alternate method I mentioned above sounds relatively smooth although does still have a few clicks and whatnot here and there, but that could be due to other reasons. Anyway, I'm going to mark this as the answer for now.

**Reply by p0p4k** [reply]
I see. Additionally you could use something like [hark](https://github.com/otalk/hark) for live streaming too. It detects silence in browser, then you can send your chunks when silence is detected or a certain time threshold has passed and get a much more real time feel.

**Reply by Disastorm** [reply]
I don't quite understand your suggestion. Why would I send chunks when silence is detected, it would be converting silence to silence? Or do you mean only send chunks when voice is detected, but it would still have the problem I mentioned above where I'd need to send 3 chunks to have the output be smooth across each chunk.

As for real time, I havn't gotten tensor working on my windows python setup so I've only been able to get it to have just under a 1 second delay at the lowest, but I'm guessing with tensor, maybe I'd be able to get like 100ms or 200ms delay which would be pretty nice. Do you have any ideas as to the performance speed increase of tensor vs cpu ?

*edit oh I think I understand your suggestion now. You'd send the data of the full voice section once silence was detected. However, that would result in the entire section/phrase/sentence not being converted until the entire sentence was over. I think that would actually sound less real time. Or is it able to actually detect little sections of silence within speech itself?

**Reply by p0p4k** [reply]
I should have worded "real-time" feel differently. My goal is slightly different from yours. I just wanted to wait till a sentence ends, detect a natural silence and then convert in that silence duration. That being said, I think your approach is good for your needs.

**Reply by Disastorm** [reply]
Ok thanks I understand. By the way, do you happen to know how the speeds compare with tensor vs non-tensor ?

**Reply by p0p4k** [reply]
Do you mean like CPU inference vs GPU inference?

**Reply by Disastorm** [reply]
oh yea sorry i meant cuda vs cpu.

**Reply by p0p4k** [reply]
Yes, in my TTS tests (not voice conversion), I do not remember the exact ratio right now, but CPU performance was very slow (around 10x) when compared to GPU.

**Reply by DanRuta** [reply]
@Disastorm I am set to try an implementation such as yous very soon. I was wondering, how long have you picked your chunks to be? And do you have any other implementation advice?

**Reply by Disastorm** [reply]
> Yes, in my TTS tests (not voice conversion), I do not remember the exact ratio right now, but CPU performance was very slow (around 10x) when compared to GPU.

oh I see thanks, thats quite a big difference.

> @Disastorm I am set to try an implementation such as yous very soon. I was wondering, how long have you picked your chunks to be? And do you have any other implementation advice?

I'm using a 300ms chunk length, although I'm using CPU tensors. With GPU maybe you can get shorter ones.
In practice my application gets something like maybe just under 1 second delay. I guess somewhere between 600ms to 1000ms between when I talk and the converted output appears. I might make a youtube video with my application in the future.

**Reply by danylo-kelvich** [reply]
Hey @Disastorm! Have you be any chance released the code somewhere? I would like to take a look at your implementation.

**Reply by Disastorm** [reply]
Sorry I havn't, and my code was pretty sloppy and confusing, too. I think there are alot of solutions for realtime ai voice cloning these days like the RVC projects, you might be able to check some of those.

**Reply by catyung** [reply]
Hi @Disastorm , thanks for your advice. I am trying to replicate your idea , in order to create low-latency TTS approach, however, I am wondering in your suggested approach, how can you locate and extract the middle chunk ? Since there is no timestamp being provided from the log. 

Appreciate your help in advance. ^^ 

Cat

**Reply by Disastorm** [reply]
@catyung I feel like the RVC stuff sounds better than yourTTS these days, but if you really still want to try this out, what I was doing is each chunk is a specific size of bytes, so you just use that to calculate it, something like
outputs = outputs[outputSize:(outputSize*2)]

**Reply by catyung** [reply]
@Disastorm Thanks ! Will try both suggestions :) ! :)

**Comment by agilebean** [comment]
Has anyone found that function for streaming CoquiTTS speech output which @Disastorm asked about?

**Reply by erogol** [reply]
this is the only streaming model https://tts.readthedocs.io/en/latest/models/xtts.html#streaming-inference

**Reply by agilebean** [reply]
@erogol that's the only thing I found too - thanks for the confirmation

**Reply by Disastorm** [reply]
I actually wasn't using a streaming model. Basically I was streaming the audio from the microphone via standard python libraries, dividing it into chunks and then processing it with yourTTS, and then merging the outputs. I guess maybe you can say it was a custom streaming function? I made a quick diagram for it just now ( using 1s chunks in the example image below ):
![OldCoquiTTS](https://github.com/coqui-ai/TTS/assets/1088694/0ad5248f-f1a7-4048-8e50-fe2129bec841)

**Comment by cod3r0k** [comment]
Do we have a stable streaming?
